src/Client/ClientBase.cpp

```diff
@@ -1542,6 +1542,14 @@ void ClientBase::processParsedSingleQuery(const String & full_query, const Strin
                     global_context->applySettingChange(change);
             }
             global_context->resetSettingsToDefaultValue(set_query->default_settings);
+
+            /// Query parameters inside SET queries should be also saved on the client side
+            ///  to override their previous definitions set with --param_* arguments
+            ///  and for substitutions to work inside INSERT ... VALUES queries
+            for (const auto & [name, value] : set_query->query_parameters)
+                query_parameters.insert_or_assign(name, value);
+
+            global_context->addQueryParameters(set_query->query_parameters);
```

Comment:
Just curious, why do we replace query_parameters for INSERT in ClientBase, for other queries like this one 
```
set param_x = 1;
SELECT * from 02476_query_parameters_insert LIMIT {x:Int32}
```
we replace in executeQuery 
'https://github.com/ClickHouse/ClickHouse/blob/2720cddeecfe0fb9517e29f64304df8d674f0886/src/Interpreters/executeQuery.cpp#L381'
is this right?

Reply:
We send data for inserts separately:
https://github.com/ClickHouse/ClickHouse/blob/b0faecf75be1e8b94b3d01ec884d1e4ddc7c730f/src/Client/ClientBase.cpp#L780-L788
It seems like it was done this way to support different optimizations for files (parallel reading), the progress bar, and maybe other stuff, but overall we parse and check those values on the client side.